src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,1006 @@ mod v1 {
     }
 }
 
+mod v2 {
+    use std::error::Error as StdError;
+    use std::io::{self, Read, Write, ErrorKind};
+    use std::fs::{self, File, OpenOptions, Metadata};
+    use std::path::{Path, PathBuf};
+    use std::sync::{Arc, RwLock};
+    use std::str;
+    use std::time::Instant;
+    use crc::crc32::{self, Digest, Hasher32};
+    use protobuf::{Message, RepeatedField};
+    use kvproto::metapb::Region;
+    use kvproto::raft_serverpb::{KeyValue, RaftSnapshotData};
+    use rocksdb::{EnvOptions, Options, SstFileWriter, IngestExternalFileOptions};
+    use storage::{CfName, ALL_CFS};
+    use util::{HandyRwLock, rocksdb};
+    use util::codec::number::{self, NumberEncoder, NumberDecoder};
+    use util::codec::bytes::{BytesEncoder, BytesDecoder};
+    use raftstore::Result as RaftStoreResult;
+
+    use super::super::keys::{enc_start_key, enc_end_key};
+    use super::super::engine::{Snapshot as DbSnapshot, Iterable};
+    use super::{SNAP_GEN_PREFIX, SNAP_REV_PREFIX, TMP_FILE_SUFFIX, SST_FILE_SUFFIX, Error, Result,
+                SnapKey, Snapshot, ApplyContext, check_abort, need_to_pack};
+
+    const META_FILE_SUFFIX: &'static str = ".meta";
+    const SNAPSHOT_META_PREFIX_SIZE: &'static str = "size";
+    const SNAPSHOT_META_PREFIX_CHECKSUM: &'static str = "checksum";
+    const ENCODE_DECODE_DESC: bool = false;
+    const DIGEST_BUFFER_SIZE: usize = 10240;
+
+    fn get_snapshot_cfs() -> Vec<CfName> {
+        let size = ALL_CFS.len();
+        let mut cfs = Vec::with_capacity(size);
+        for cf in ALL_CFS {
+            if !need_to_pack(cf) {
+                continue;
+            }
+            cfs.push(*cf);
+        }
+        cfs
+    }
+
+    fn get_file_size(path: &str) -> io::Result<u64> {
+        let meta = try!(fs::metadata(path));
+        Ok(meta.len())
+    }
+
+    fn file_exists(file: &str) -> bool {
+        let path = Path::new(file);
+        path.exists() && path.is_file()
+    }
+
+    fn delete_file(file: &str) -> bool {
+        if let Err(e) = fs::remove_file(file) {
+            warn!("failed to delete file {}: {:?}", file, e);
+            return false;
+        }
+        true
+    }
+
+    fn calc_checksum(p: &str) -> io::Result<u32> {
+        let mut digest = Digest::new(crc32::IEEE);
+        let mut f = try!(OpenOptions::new().read(true).open(&p));
+        let mut buf = vec![0; DIGEST_BUFFER_SIZE];
+        loop {
+            let n = try!(f.read(&mut buf[..]));
+            digest.write(&buf[..n]);
+            if n < DIGEST_BUFFER_SIZE {
+                return Ok(digest.sum32());
+            }
+        }
+    }
+
+    fn encode_cf_size_checksums(cf_files: &[CfFile]) -> RaftStoreResult<(u64, Vec<KeyValue>)> {
```

Comment:
Adding a json format doesn't reduce the work. It still requires to validate the filed name and value and require the overhead of encoding/decoding itself. Since the meta is encoded inside a protobuf format, the readability of json will be lost.